bfasst/transform/spydrnet_injector.py

```diff
@@ -0,0 +1,199 @@
+            )
+        else:
+            config_as_binary = (
+                config_as_binary[:bit_number] + "1" + config_as_binary[bit_number + 1 :]
+            )
+
+        config_as_hex = hex(int(config_as_binary, 2))
+        lut_properties["INIT"] = (
+            config_string_prefixed.split("H")[0] + "h" + str(config_as_hex).upper()[2:]
+        )
+
+    def compose_corrupt_netlist(self):
+        """Writes the netlist to the corrupted netlist path in the design"""
+        sdn.compose(self.clean_netlist, self.design.flow_paths["corrupted_netlist_path"])
+
+    def inject_wire_swap(self):
+        """Injects a wire swap error into the netlist"""
+        three_instances = self.get_random_instances(3)
+        first_instance_pins = self.get_outer_pin_inputs(three_instances[0])
+
+        selected_input = first_instance_pins[randrange(len(first_instance_pins))]
+        driving_pin = self.get_source(selected_input)
+        new_driver = self.get_new_driver(driving_pin, three_instances[1], three_instances[2])
+
+        self.detach_wire(selected_input)
+        self.attach_wire(selected_input, new_driver)
+
+        self.compose_corrupt_netlist()
+
+    def get_random_instances(self, num_instances):
+        """Gets a list of random instances from the netlist"""
+        instances = [
+            instance
+            for instance in self.clean_netlist.get_instances()
+            if "GND" not in instance.reference.name.upper()
+            and "VCC" not in instance.reference.name.upper()
+            and "VDD" not in instance.reference.name.upper()
+        ]
+        return sample(instances, num_instances)
+
+    def get_outer_pin_inputs(self, instance):
+        """Gets all the outer pins of the given instance that are inputs"""
+        outer_pin_inputs = [pin for pin in instance.pins if pin.inner_pin.port.direction is sdn.IN]
+        if not outer_pin_inputs:
+            outer_pin_inputs = self.get_unisim_outer_pin_inputs(instance)
+
+        return outer_pin_inputs
+
+    def get_unisim_outer_pin_inputs(self, instance):
+        """Collect all the pins for a UNISIM cell that are inputs"""
+
+        cell_inputs = (
+            (("LUT6_2",), ("I0", "I1", "I2", "I3", "I4", "I5")),
+            (("IBUF", "OBUF", "OBUFT"), ("I", "T")),
+            (("GND",), ()),
+            (("VCC",), ()),
+            (("FDSE", "FDRE"), ("D", "CE", "R", "C", "S")),
+            (("CARRY4",), ("CI", "CYINIT", "DI", "S")),
+            (("BUFGCTRL",), ("CE0", "CE1", "I0", "I1", "IGNORE0", "IGNORE1", "S0", "S1")),
+            (("MUXF7", "MUXF8"), ("I0", "I1", "S")),
+        )
+
+        outer_pin_inputs = []
+        for name, inputs in cell_inputs:
+            if instance.reference.name in name:
+                pins = list(instance.pins)
+                for pin in pins:
+                    if pin.inner_pin.port.name in inputs:
+                        outer_pin_inputs.append(pin)
+
+        return outer_pin_inputs
```

Comment:
Is this duplicated from the structural code?  Would be good to move to a common location.

Reply:
Is the utils file an okay place to abstract this to?

Reply:
How about `bfasst/vendor_data/xilinx.py`

Reply:
Sounds good

Reply:
Done